### ISSUE TITLE
[java] JUnit4TestShouldUseTestAnnotation

### ISSUE THREAD

**Issue by vanniktech** [pmd/pmd#375]
**Rule Set:**

```xml
<rule ref="rulesets/java/migrating.xml"/>
```
The following code:

```java
import junit.framework.TestCase;

public class Test1234 extends TestCase {
  public void test12() {

  }
}
```

is flagged by PMD:

```xml
<violation beginline="13" endline="15" begincolumn="3" endcolumn="3" rule="JUnit4TestShouldUseTestAnnotation" ruleset="Migration" package="default" class="Test1234" externalInfoUrl="https://pmd.github.io/pmd-5.5.5/pmd-java/rules/java/migrating.html#JUnit4TestShouldUseTestAnnotation" priority="3">
JUnit 4 tests that execute tests should use the @Test annotation
</violation>
```

I'm a bit confused since the class is using JUnit 3 and not annotating it with `@Test` is perfectly valid and normal. Is this a false positive?

**Running PMD 5.5.5 through:** *[Gradle]*

**Comment by ryan-gustafson** [comment]
This is a migration rule for moving from from JUnit 3 to 4, in which case `@Test` would be needed.  If you're not intending to use/migrate to JUnit 4, then don't use the rule.

If you want some of the migration rules, but not others, either explicitly include those rules, or include the full ruleset and exclude those you don't want.  See:  https://pmd.github.io/pmd-5.5.5/customizing/howtomakearuleset.html

Closing this issue, as I believe this is working as intended.

**Comment by vanniktech** [comment]
I get that although I'm forced to use Junit3 and can't use Junit 4 (the framework that I want to test only exposes Junit3 compatibility). I'd expect this to not flag since I'm not using Junit4 and I'm explicitly stating that I'm using Junit3 by extending from `TestCase`.

**Comment by ryan-gustafson** [comment]
I understand the Rule is not disabling itself when someone is using JUnit 3, but I think the case where one would use this rule is when one is in the process of transitioning between 3 and 4 and wants to know what to fix to help complete the migration.  Once on JUnit 4 the rule can stay too, to make sure you don't miss `@Test` if you still use the test prefix naming convention (I personally use it).

Since you are not transitioning your code from 3 to 4, you are not migrating, and should not be using this rule.  Blanket application of the `migration` ruleset to arbitrary code seems inappropriate to me.

I suppose there are cases where one could be in an interim migration, but I would use other means to explicitly avoid code which is intended to stay on JUnit 3, versus code which is JUnit 3 that should be moved to JUnit 4.  Ideally one would try to avoid a state of duality, and just migrate to JUnit 4 outright.

**Comment by vanniktech** [comment]
Alright for now I just excluded it and PMD is no longer complaining. Thanks :)
